pkg/apiserver/user/sqlauth/sqlauth.go

```diff
@@ -50,6 +79,64 @@ func (a *Authenticator) Authenticate(f user.AuthenticateForm) (*utils.SessionUse
 	}
 	defer utils.CloseTiDBConnection(db) //nolint:errcheck
 
+	// Check privileges
+	// 1. Check whether TiDB SEM is enabled
+	resData, err := a.tidbClient.SendGetRequest("/config")
+	if err != nil {
+		return nil, user.ErrSignInOther.WrapWithNoMessage(err)
+	}
+	var config tidbSecurityConfig
+	err = json.Unmarshal(resData, &config)
+	if err != nil {
+		return nil, user.ErrSignInOther.WrapWithNoMessage(err)
+	}
+	// 2. Get grants
+	var grantRows []string
+	err = db.Raw("show grants for current_user()").Find(&grantRows).Error
+	if err != nil {
+		return nil, user.ErrSignInOther.WrapWithNoMessage(err)
+	}
+	grants := parseGrants(grantRows)
+	// 3. Check
+	// Following base privileges are required
+	// - ALL PRIVILEGES
+	// - or
+	// - PROCESS
+	// - SHOW DATABASES
+	// - SYSTEM_VARIABLES_ADMIN or SUPER (SUPER includes SYSTEM_VARIABLES_ADMIN)
+	// - CONFIG
+	// - DASHBOARD_CLIENT
+	// When TiDB SEM is enabled, following extra privileges are required
+	// - RESTRICTED_TABLES_ADMIN
+	// - RESTRICTED_STATUS_ADMIN
+	basePrivileges := 0
+	extraPrivileges := 0
+	for _, priv := range grants {
+		switch priv {
+		case privProcess:
+			basePrivileges = basePrivileges | 1
+		case privShowDB:
+			basePrivileges = basePrivileges | (1 << 1)
+		case privSystemVarAdmin:
+		case privSuper:
+			basePrivileges = basePrivileges | (1 << 2)
+		case privConfig:
+			basePrivileges = basePrivileges | (1 << 3)
+		case privDashboardClient:
+			basePrivileges = basePrivileges | (1 << 4)
+		case privAll:
+			basePrivileges = 0b11111
```

Comment:
How about just using:

```
if hasPriv(ALL PRIVILEGES) {
  pass
} else {
  if !hasPriv(PROCESS) {
    not pass
  }
  if ... {
    ...
  }
}
```

The golden rule to check whether your code is general enough or provides enough flexibility is to try to add something new to it, or removing something from it. For example, suppose we want a new privilege FOO to be checked in addition to current privileges, how many code do we need to change? how is the likelihood that something is forget to change? will we easily discover that something is forgotten?

By using this rule, it can be easily seen that:
1. During adding a new privilege, `privAll` is something easily to forget, and it must be changed accordingly, leaving a mistake possibility.
2. During removing a privilege, similar issues occurs. And what's worse, the bit shift order 1, 2, 3, 4, 5 needs to be changed.

A more flexible implementation should provide less and more integrated change for these future change scenarios. And as you can see, the simple `if else` works just better.